### DISCUSSION TITLE
Clearing one task instance will execute all task instances that have not previously run

### DISCUSSION THREAD

**Discussion by edwardwang888** [apache/airflow#18347] [General]
### Apache Airflow version

main (development)

### Operating System

OS X (development environment)

### Deployment

Other

### What happened

1. Here is a DAG with two tasks that are newly added. These task instances have not been run yet.
![image](https://user-images.githubusercontent.com/40527812/133390015-b2d66698-992a-4af8-a936-1cdf9ae11fc9.png)

1. Clear only the first two task instances.
![image](https://user-images.githubusercontent.com/40527812/133389044-548ea2db-b21a-4e13-97b5-c19745d489a6.png)

1. Tree view after clearing.
![image](https://user-images.githubusercontent.com/40527812/133389169-60c5eeed-ff45-4507-b091-5547e090bffe.png)

1. Even though only two task instances were cleared, all four task instances are executed.
![image](https://user-images.githubusercontent.com/40527812/133389748-7b8ee781-b9ba-4e18-a21c-2ba52373376e.png)

### What you expected to happen

Only the task instances that were cleared should execute.

### How to reproduce

Add tasks to a DAG, and then clear any task instance other than the ones that were just added. Another way is to add tasks to a DAG that is currently running - the newly added task instances will execute (in whichever DAG runs are running) once added.

### Anything else

Based on my observations, I believe this happens because the newly-added task instances do not yet exist in the `task_instance` table, so the scheduler(?) adds them to the table with state = `NULL` and they get executed. Perhaps this could be solved if we only add the task instances to the table if that DAG run does not have any existing task instances (new DAG run) or if the DAG run is cleared.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
I think the behaviour is as expected currently - until we have DAG versioning implemented https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-36+DAG+Versioning 

Currently Airflow does not support DAG evolution. until versioning is implemented, the best way to add substantial changes to a DAG is to copy it and have a new dag_id.

**Reply by potiuk** [reply]
I converted it to a discussion, as someone might want to add something to it.
